packages/typeahead/src/react/index.js

```diff
@@ -0,0 +1,257 @@
+  const { activeValue, suggestions, ...rest } = props
+
+  const items = React.useMemo(() => {
+    if (suggestions.length <= 0)
+      return (
+        <ActionMenu.Item disabled value="">
+          no results found
+        </ActionMenu.Item>
+      )
+
+    return suggestions.map(sug => (
+      <ActionMenu.Item
+        key={sug.index}
+        icon={<CheckIcon visible={sug.value === activeValue} />}
+        value={sug.value}
+      >
+        {sug.comp}
+      </ActionMenu.Item>
+    ))
+  }, [activeValue, suggestions])
+
+  return (
+    <ActionMenu
+      {...rest}
+      ref={forwardedRef}
+      origin={ActionMenu.origins.topLeft}
+      shouldFocusOnMount={false}
+    >
+      {items}
+    </ActionMenu>
+  )
+})
+
+SuggestionsMenu.propTypes = {
+  activeValue: PropTypes.string.isRequired,
+  onChange: PropTypes.func,
+  suggestions: PropTypes.arrayOf(
+    PropTypes.shape({
+      comp: PropTypes.node.isRequired,
+      index: PropTypes.number.isRequired,
+      value: PropTypes.string.isRequired
+    })
+  ).isRequired
+}
+
+const Suggestion = React.forwardRef((props, forwardedRef) => {
+  return <div ref={forwardedRef} {...props} />
+})
+
+Suggestion.propTypes = {
+  children: PropTypes.string.isRequired
+}
+
+Typeahead.propTypes = {
+  children: PropTypes.oneOfType([
+    elementOfType(Suggestion),
+    PropTypes.arrayOf(elementOfType(Suggestion))
+  ]),
+  filterFn: PropTypes.func,
+  loading: PropTypes.bool,
+  onBlur: PropTypes.func,
+  onChange: PropTypes.func,
+  onFocus: PropTypes.func,
+  value: PropTypes.string,
+
+  // props used by underlying TextInput
+  appearance: PropTypes.any,
+  disabled: PropTypes.any,
+  error: PropTypes.any,
+  label: PropTypes.any,
+  name: PropTypes.any,
+  placeholder: PropTypes.any,
+  subLabel: PropTypes.any
+}
+Typeahead.defaultProps = {
+  filterFn: filterSuggestions
+}
+
+Typeahead.appearances = vars.appearances
+
+Typeahead.Suggestion = Suggestion
+
+export const appearances = vars.appearances
+
+export default Typeahead
+
+function getSuggestionValue(suggestionInst) {
+  return suggestionInst.props.children
+}
+
+function filterSuggestions(searchTerm, children) {
+  if (!searchTerm || searchTerm.length <= 1) return children
+
+  const term = searchTerm.toLowerCase()
+  const matches = ({ value }) => value.toLowerCase().includes(term)
+
+  return children.filter(matches)
+}
+
+function useOnDocumentClick(handler) {
+  React.useEffect(() => {
+    document.addEventListener('mousedown', handler)
+    document.addEventListener('touchstart', handler)
+
+    return () => {
+      document.removeEventListener('mousedown', handler)
+      document.removeEventListener('touchstart', handler)
+    }
+  }, [handler])
+}
+
+function usePortal() {
+  const ref = React.useRef(document.createElement('div'))
+
+  React.useEffect(() => {
+    const { current } = ref
+    document.body.appendChild(current)
+
+    return () => {
+      document.body.removeChild(current)
+    }
+  }, [])
+
+  return ref
+}
```

Comment:
This file is potentially more readable if it's broken up into other files, like done with ActionMenu recently.  

That file was probably bigger.  But, navigation and structure discovery was helped by breaking it up.  But once I started it, I had a hard time stopping breaking things out. :)

Reply:
agreed...i'll refactor